src/main/java/org/broadinstitute/hellbender/utils/locusiterator/LocusIteratorByState.java

```diff
@@ -0,0 +1,348 @@
+            final AlignmentContext context = next();
+
+            if ( context == null ){
+                // we ran out of data
+                return null;
+            }
+
+            if ( context.getPosition() == position ) {
+                return context;
+            }
+
+            if ( context.getPosition() > position) {
+                return stopAtFirstNonEmptySiteAfterPosition ? context : null;
+            }
+        }
+
+        return null;
+    }
+
+    /**
+     * Creates the next alignment context from the given state.  Note that this is implemented as a
+     * lazy load method. nextAlignmentContext MUST BE null in order for this method to advance to the
+     * next entry.
+     */
+    private void lazyLoadNextAlignmentContext() {
+        while (nextAlignmentContext == null && readStates.hasNext()) {
+            readStates.collectPendingReads();
+
+            final Locatable location = getLocation();
+            final Map<String, ReadPileup> fullPileupPerSample = new HashMap<>();
+
+            for (final Map.Entry<String, PerSampleReadStateManager> sampleStatePair : readStates ) {
+                final String sample = sampleStatePair.getKey();
+                final PerSampleReadStateManager readState = sampleStatePair.getValue();
+                final Iterator<AlignmentStateMachine> iterator = readState.iterator();
+                final List<PileupElement> pile = new ArrayList<>(readState.size());
+
+                while (iterator.hasNext()) {
+                    // state object with the read/offset information
+                    final AlignmentStateMachine state = iterator.next();
+                    final GATKRead read = state.getRead();
+                    final CigarOperator op = state.getCigarOperator();
+
+                    if (op == CigarOperator.N){ // N's are never added to any pileup
+                        continue;
+                    }
+
+                    if (!dontIncludeReadInPileup(read, location.getStart())) {
+                        if ( ! includeReadsWithDeletionAtLoci && op == CigarOperator.D ) {
+                            continue;
+                        }
+
+                        pile.add(state.makePileupElement());
+                    }
+                }
+
+                if (! pile.isEmpty() ){ // if this pileup added at least one base, add it to the full pileup
+                    fullPileupPerSample.put(sample, new ReadPileup(location, pile));
+                }
+            }
+
+            readStates.updateReadStates(); // critical - must be called after we get the current state offsets and location
+            if (!fullPileupPerSample.isEmpty()){ // if we got reads with non-D/N over the current position, we are done
+                if (fullPileupPerSample.keySet().size() > 1){
+                    throw new UnsupportedOperationException("Multiple samples are currently not supported in GATK4. Samples here:" + fullPileupPerSample.keySet());
```

Comment:
There's no reason to throw if we have multi-sample data, as far as I can tell, as you've ported all the per-sample tracking -- remove this restriction, as we want to support multi-sample data for https://github.com/broadinstitute/gatk/issues/1464